### DISCUSSION TITLE
build CREXX with Ninja

### DISCUSSION THREAD

**Discussion by rvjansen** [adesutherland/CREXX#235] [Show and tell]
I always wondered why a C compile is so slow, when compared to e.g. Java. Now I know where a part of that goes.
Entirely by happenstance - well trying to solve something nasty actually - I was building SWI Prolog from source on my M1 Mini. This happens to use the Ninja build system, and to remain with the Zeitgeist: you would not believe what happened next.

I kept half an eye on my screen and half an eye on my brie sandwich, and I saw the build fail, at least it stoppped, and I thought I knew I had to spend some time hunting down dependencies.
But lo and behold, it had built correctly and also ran correctly. I noticed I just installed something called Ninja from homebrew, following the instructions, and when zooming in a bit, I learned that it is a build system.

To make a short story much shorter, a normal make of CREXX takes almost a minute (56 sec elapsed) on this Mac mini. With Ninja, it takes 15 seconds elapsed, but uses much more CPU (539% to be exact). I only care about the elapsed time here.

To use ninja, install it (homebrew has it, brew install ninja, and for Ubuntu: sudo apt install ninja-build), remove your CmakeCache.txt in your existing build directory, and run:
```
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ../CREXX
```
followed by:
```
ninja
```
It turns out that cmake knows how to make ninja files. 

On the raspi 4, a complete (across-the-ocean) build goes from 7 minutes to 3 minutes. On the X86_64 NUC,  Ninja take 38 seconds as opposed to make's 1 minute and 33s.

**Comment by Peter-Jacob** [comment]
I am not surprised: Ninja's methods included espionage, sabotage, deception, assassination, and surprise attacks. Their covert methods of waging irregular warfare were deemed dishonorable and beneath the honor of the samurai.
I learned 3 years of Nin-Jutsu under a high-rated sensei, there I learned: if force is not sufficient use brute force! 
Any question?

**Reply by rvjansen** [reply]
So REALLY EVERYBODY was Kung-Fu fighting?

**Comment by mgrossmann** [comment]
🤓😂🙏

René Vincent Jansen ***@***.***> schrieb am Mo. 29. Nov. 2021
um 22:59:

> So REALLY EVERYBODY was Kung-Fu fighting?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adesutherland/CREXX/discussions/235#discussioncomment-1718764>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF7Q7OMQFSRA7K5TLO2LSLUOPZVPANCNFSM5I7TAEEA>
> .
>
-- 
Von Gmail Mobile gesendet
